PWGHF/HFC/TableProducer/correlatorDsHadrons.cxx

```diff
@@ -582,68 +570,80 @@ struct HfCorrelatorDsHadrons {
   }
   PROCESS_SWITCH(HfCorrelatorDsHadrons, processMcRec, "Process MC Reco mode", false);
 
+  Preslice<CandDsMcGen> perCollisionCandMc = o2::aod::mcparticle::mcCollisionId;
+
   /// Ds-Hadron correlation pair builder - for MC gen-level analysis (no filter/selection, only true signal)
-  void processMcGen(SelCollisionsWithDsMc::iterator const& mcCollision,
+  void processMcGen(SelCollisionsWithDsWithMc const& collisions,
+                    soa::Join<aod::McCollisions, aod::MultsExtraMC> const&,
                     CandDsMcGen const& mcParticles)
   {
     BinningTypeMcGen corrBinningMcGen{{zPoolBins, multPoolBins}, true};
-    int poolBin = corrBinningMcGen.getBin(std::make_tuple(mcCollision.posZ(), mcCollision.multMCFT0A()));
-    registry.fill(HIST("hCollisionPoolBin"), poolBin);
-    registry.fill(HIST("hMultFT0AMcGen"), mcCollision.multMCFT0A());
-    bool isDsPrompt = false;
-    bool isDecayChan = false;
-    int trackOrigin = -1;
-    // MC gen level
-    for (const auto& particle : mcParticles) {
-      // check if the particle is Ds
-      if (std::abs(particle.pdgCode()) != Pdg::kDS) {
-        continue;
-      }
-      if ((std::abs(particle.flagMcMatchGen()) == 1 << aod::hf_cand_3prong::DecayType::DsToKKPi) && (particle.flagMcDecayChanGen() == decayChannel)) {
-        double yD = RecoDecay::y(particle.pVector(), MassDS);
-        if (std::abs(yD) > yCandGenMax || particle.pt() < ptCandMin || particle.pt() > ptCandMax) {
+
+    for (const auto& collision : collisions) {
```

Comment:
I am not sure about your logic; if you have two reconstructed collisions that point to the same generated collision (split vertex), then you consider twice the same generated particles. Is this really what you want to do? Wouldn't it be better to loop over mcCollisions and check if they have at least a reconstructed collision associated to it?